File: space-game/Core.Tests/Data/EnumerableExtensionsTestsData.vb
Imports Core.Classes
Imports Core.Enums
Imports NUnit.Framework

Namespace Data
    Public Class EnumerableExtensionsTestsData
        Public Shared Function HasAny_WhenGivenAList_ReturnsExpected_Data() As List(Of TestCaseData)
            Dim nullList As List(Of String) = Nothing
            Dim emptyList As New List(Of String)
            Dim listWithMembers = New List(Of String) From {"a", "b"}

            Return New List(Of TestCaseData) From {
                New TestCaseData(nullList, False),
                New TestCaseData(emptyList, False),
                New TestCaseData(listWithMembers, True)
            }
        End Function

        Private Shared _samples As SampleTextures

        Private Shared ReadOnly Property Samples As SampleTextures
            Get
                If _samples Is Nothing Then
                    _samples = New SampleTextures(TextureType.Diffuse)
                End If
                Return _samples
            End Get
        End Property

        Private Shared ReadOnly All As New List(Of Texture) From {Samples.Low,
                                                                    Samples.Medium,
                                                                    Samples.High}

        Private Shared ReadOnly MediumHigh As New List(Of Texture) From {Samples.Medium,
                                                                            Samples.High}

        Private Shared ReadOnly MediumLow As New List(Of Texture) From {Samples.Medium,
                                                                            Samples.Low}

        Private Shared ReadOnly LowHigh As New List(Of Texture) From {Samples.High,
                                                                            Samples.Low}

        Private Shared ReadOnly OnlyLow As New List(Of Texture) From {Samples.Low}

        Private Shared ReadOnly None As New List(Of Texture)

        Private Shared ReadOnly AllWithLowValue As New List(Of Texture) From {Samples.Low,
                                                                                Samples.EmptyMedium,
                                                                                Samples.EmptyHigh}

        Private Shared ReadOnly AllWithHighalue As New List(Of Texture) From {Samples.EmptyLow,
                                                                                Samples.EmptyMedium,
                                                                                Samples.High}

        Public Shared Function GetHighestAvailableResolution_WhenGivenAList_ReturnsExpected_Data() As List(Of TestCaseData)
            Return New List(Of TestCaseData) From {
                New TestCaseData(All, Samples.Type, Samples.High.Path),
                New TestCaseData(MediumHigh, Samples.Type, Samples.High.Path),
                New TestCaseData(OnlyLow, Samples.Type, Samples.Low.Path),
                New TestCaseData(MediumLow, Samples.Type, Samples.Medium.Path),
                New TestCaseData(LowHigh, Samples.Type, Samples.High.Path),
                New TestCaseData(None, Samples.Type, ""),
                New TestCaseData(All, Samples.Type + 1, ""),
                New TestCaseData(AllWithLowValue, Samples.Type, Samples.Low.Path),
                New TestCaseData(AllWithHighalue, Samples.Type, Samples.High.Path)
            }
        End Function

        Public Shared Function GetLowestAvailableResolution_WhenGivenAList_ReturnsExpected_Data() As List(Of TestCaseData)
            Return New List(Of TestCaseData) From {
                New TestCaseData(All, Samples.Type, Samples.Low.Path),
                New TestCaseData(MediumHigh, Samples.Type, Samples.Medium.Path),
                New TestCaseData(OnlyLow, Samples.Type, Samples.Low.Path),
                New TestCaseData(MediumLow, Samples.Type, Samples.Low.Path),
                New TestCaseData(LowHigh, Samples.Type, Samples.Low.Path),
                New TestCaseData(None, Samples.Type, ""),
                New TestCaseData(All, Samples.Type + 1, ""),
                New TestCaseData(AllWithLowValue, Samples.Type, Samples.Low.Path),
                New TestCaseData(AllWithHighalue, Samples.Type, Samples.High.Path)
            }
        End Function

        Public Shared Function GetHighestAvailableResolutionForEachType_High() As List(Of TestCaseData)
            Dim diffuseSamples = New SampleTextures(TextureType.Diffuse)
            Dim bumpSamples = New SampleTextures(TextureType.Bump)
            Dim emissiveSamples = New SampleTextures(TextureType.Emissive)

            Dim textures As New List(Of Texture)
            textures.AddRange({diffuseSamples.Low, diffuseSamples.Medium, diffuseSamples.High})
            textures.AddRange({bumpSamples.Low, bumpSamples.Medium, bumpSamples.High})
            textures.AddRange({emissiveSamples.Low, emissiveSamples.Medium, emissiveSamples.High})

            Dim expected As New List(Of Texture) From {
                    diffuseSamples.High,
                    bumpSamples.High,
                    emissiveSamples.High
                }

            Return New List(Of TestCaseData) From {
                    New TestCaseData(textures, expected)
                }
        End Function

        Public Shared Function GetHighestAvailableResolutionForEachType_Medium() As List(Of TestCaseData)
            Dim diffuseSamples = New SampleTextures(TextureType.Diffuse)
            Dim bumpSamples = New SampleTextures(TextureType.Bump)
            Dim emissiveSamples = New SampleTextures(TextureType.Emissive)

            Dim textures As New List(Of Texture)
            textures.AddRange({diffuseSamples.Low, diffuseSamples.Medium})
            textures.AddRange({bumpSamples.Low, bumpSamples.Medium})
            textures.AddRange({emissiveSamples.Low, emissiveSamples.Medium})

            Dim expected As New List(Of Texture) From {
                    diffuseSamples.Medium,
                    bumpSamples.Medium,
                    emissiveSamples.Medium
                }

            Return New List(Of TestCaseData) From {
                    New TestCaseData(textures, expected)
                }
        End Function

        Public Shared Function GetHighestAvailableResolutionForEachType_Low() As List(Of TestCaseData)
            Dim diffuseSamples = New SampleTextures(TextureType.Diffuse)
            Dim bumpSamples = New SampleTextures(TextureType.Bump)
            Dim emissiveSamples = New SampleTextures(TextureType.Emissive)

            Dim textures As New List(Of Texture)
            textures.AddRange({diffuseSamples.Low})
            textures.AddRange({bumpSamples.Low})
            textures.AddRange({emissiveSamples.Low})

            Dim expected As New List(Of Texture) From {
                    diffuseSamples.Low,
                    bumpSamples.Low,
                    emissiveSamples.Low
                }

            Return New List(Of TestCaseData) From {
                    New TestCaseData(textures, expected)
                }
        End Function

        Public Shared Function GetHighestAvailableResolutionForEachType_Mix() As List(Of TestCaseData)
            Dim diffuseSamples = New SampleTextures(TextureType.Diffuse)
            Dim bumpSamples = New SampleTextures(TextureType.Bump)
            Dim emissiveSamples = New SampleTextures(TextureType.Emissive)

            Dim textures As New List(Of Texture)
            textures.AddRange({diffuseSamples.Low, diffuseSamples.Medium, diffuseSamples.High})
            textures.AddRange({bumpSamples.Low, bumpSamples.Medium})
            textures.AddRange({emissiveSamples.Low})

            Dim expected As New List(Of Texture) From {
                    diffuseSamples.High,
                    bumpSamples.Medium,
                    emissiveSamples.Low
                }

            Return New List(Of TestCaseData) From {
                    New TestCaseData(textures, expected)
                }
        End Function

        Public Shared Function GetHighestAvailableResolutionForEachType_Empty() As List(Of TestCaseData)
            Dim diffuseSamples = New SampleTextures(TextureType.Diffuse)
            Dim bumpSamples = New SampleTextures(TextureType.Bump)
            Dim emissiveSamples = New SampleTextures(TextureType.Emissive)

            Dim textures As New List(Of Texture)
            diffuseSamples.High.Path = ""
            bumpSamples.Medium.Path = ""
            emissiveSamples.Low.Path = ""

            textures.AddRange({diffuseSamples.Low, diffuseSamples.Medium, diffuseSamples.High})
            textures.AddRange({bumpSamples.Low, bumpSamples.Medium})
            textures.AddRange({emissiveSamples.Low})

            Dim expected As New List(Of Texture) From {
                    diffuseSamples.Medium,
                    bumpSamples.Low
                }

            Return New List(Of TestCaseData) From {
                    New TestCaseData(textures, expected)
                }
        End Function

        Public Shared Function IsEquivalent_WhenGivenAList_ReturnsExpected_NullAndEmpty() As List(Of TestCaseData)
            Return New List(Of TestCaseData) From {
                    New TestCaseData(New List(Of Texture), Nothing, False)
                  }
        End Function

        Public Shared Function IsEquivalent_WhenGivenAList_ReturnsExpected_Contents() As List(Of TestCaseData)

            Dim allCopy = New List(Of Texture)(All)
            Dim allWithExtra = New List(Of Texture)(All)

            allWithExtra.Add(New Texture() With {.QualityEnum = TextureQuality.Medium,
                                            .TypeEnum = TextureType.Opacity,
                                            .Path = "extra texture"})

            All.Reverse()

            Return New List(Of TestCaseData) From {
                    New TestCaseData(All, All, True),
                    New TestCaseData(allCopy, allCopy, True),
                    New TestCaseData(allCopy, All, True),
                    New TestCaseData(All, OnlyLow, False),
                    New TestCaseData(All, allWithExtra, False)
                  }
        End Function
    End Class
End Namespace

File: space-game/Core.Tests/Data/EnumerableExtensionsTestsData.cs
using Core.Classes;
using Core.Enums;
using NUnit.Framework;
using System.Collections.Generic;

namespace Core.Tests.Data
{
    public class EnumerableExtensionsTestsData
    {
        public static List<TestCaseData> HasAny_WhenGivenAList_ReturnsExpected_Data()
        {
            List<string> nullList = null;
            List<string> emptyList = new List<string>();
            List<string> listWithMembers = new List<string> {
                "a",
                "b"
            };

            return new List<TestCaseData> {
                new TestCaseData(nullList, false),
                new TestCaseData(emptyList, false),
                new TestCaseData(listWithMembers, true)
            };
        }

        private static SampleTextures _samples;
        private static SampleTextures Samples
        {
            get
            {
                if (_samples == null)
                {
                    _samples = new SampleTextures(TextureType.Diffuse);
                }
                return _samples;
            }
        }

        private static readonly List<Texture> All = new List<Texture> {
            Samples.Low,
            Samples.Medium,
            Samples.High
        };
        private static readonly List<Texture> MediumHigh = new List<Texture> {
            Samples.Medium,
            Samples.High
        };
        private static readonly List<Texture> MediumLow = new List<Texture> {
            Samples.Medium,
            Samples.Low
        };
        private static readonly List<Texture> LowHigh = new List<Texture> {
            Samples.High,
            Samples.Low
        };

        private static readonly List<Texture> OnlyLow = new List<Texture> { Samples.Low };

        private static readonly List<Texture> None = new List<Texture>();
        private static readonly List<Texture> AllWithLowValue = new List<Texture> {
            Samples.Low,
            Samples.EmptyMedium,
            Samples.EmptyHigh
        };
        private static readonly List<Texture> AllWithHighalue = new List<Texture> {
            Samples.EmptyLow,
            Samples.EmptyMedium,
            Samples.High
        };
        public static List<TestCaseData> GetHighestAvailableResolution_WhenGivenAList_ReturnsExpected_Data()
        {
            return new List<TestCaseData> {
                new TestCaseData(All, Samples.Type, Samples.High.Path),
                new TestCaseData(MediumHigh, Samples.Type, Samples.High.Path),
                new TestCaseData(OnlyLow, Samples.Type, Samples.Low.Path),
                new TestCaseData(MediumLow, Samples.Type, Samples.Medium.Path),
                new TestCaseData(LowHigh, Samples.Type, Samples.High.Path),
                new TestCaseData(None, Samples.Type, ""),
                new TestCaseData(All, Samples.Type + 1, ""),
                new TestCaseData(AllWithLowValue, Samples.Type, Samples.Low.Path),
                new TestCaseData(AllWithHighalue, Samples.Type, Samples.High.Path)
            };
        }

        public static List<TestCaseData> GetLowestAvailableResolution_WhenGivenAList_ReturnsExpected_Data()
        {
            return new List<TestCaseData> {
                new TestCaseData(All, Samples.Type, Samples.Low.Path),
                new TestCaseData(MediumHigh, Samples.Type, Samples.Medium.Path),
                new TestCaseData(OnlyLow, Samples.Type, Samples.Low.Path),
                new TestCaseData(MediumLow, Samples.Type, Samples.Low.Path),
                new TestCaseData(LowHigh, Samples.Type, Samples.Low.Path),
                new TestCaseData(None, Samples.Type, ""),
                new TestCaseData(All, Samples.Type + 1, ""),
                new TestCaseData(AllWithLowValue, Samples.Type, Samples.Low.Path),
                new TestCaseData(AllWithHighalue, Samples.Type, Samples.High.Path)
            };
        }

        public static List<TestCaseData> GetHighestAvailableResolutionForEachType_High()
        {
            SampleTextures diffuseSamples = new SampleTextures(TextureType.Diffuse);
            SampleTextures bumpSamples = new SampleTextures(TextureType.Bump);
            SampleTextures emissiveSamples = new SampleTextures(TextureType.Emissive);

            List<Texture> textures = new List<Texture>();
            textures.AddRange(new List<Texture> {
                diffuseSamples.Low,
                diffuseSamples.Medium,
                diffuseSamples.High
            });
            textures.AddRange(new List<Texture> {
                bumpSamples.Low,
                bumpSamples.Medium,
                bumpSamples.High
            });
            textures.AddRange(new List<Texture> {
                emissiveSamples.Low,
                emissiveSamples.Medium,
                emissiveSamples.High
            });

            List<Texture> expected = new List<Texture> {
                diffuseSamples.High,
                bumpSamples.High,
                emissiveSamples.High
            };

            return new List<TestCaseData> { new TestCaseData(textures, expected) };
        }

        public static List<TestCaseData> GetHighestAvailableResolutionForEachType_Medium()
        {
            SampleTextures diffuseSamples = new SampleTextures(TextureType.Diffuse);
            SampleTextures bumpSamples = new SampleTextures(TextureType.Bump);
            SampleTextures emissiveSamples = new SampleTextures(TextureType.Emissive);

            List<Texture> textures = new List<Texture>();
            textures.AddRange(new List<Texture> {
                diffuseSamples.Low,
                diffuseSamples.Medium
            });
            textures.AddRange(new List<Texture> {
                bumpSamples.Low,
                bumpSamples.Medium
            });
            textures.AddRange(new List<Texture> {
                emissiveSamples.Low,
                emissiveSamples.Medium
            });

            List<Texture> expected = new List<Texture> {
                diffuseSamples.Medium,
                bumpSamples.Medium,
                emissiveSamples.Medium
            };

            return new List<TestCaseData> { new TestCaseData(textures, expected) };
        }

        public static List<TestCaseData> GetHighestAvailableResolutionForEachType_Low()
        {
            SampleTextures diffuseSamples = new SampleTextures(TextureType.Diffuse);
            SampleTextures bumpSamples = new SampleTextures(TextureType.Bump);
            SampleTextures emissiveSamples = new SampleTextures(TextureType.Emissive);

            List<Texture> textures = new List<Texture>();
            textures.Add(diffuseSamples.Low);
            textures.Add(bumpSamples.Low);
            textures.Add(emissiveSamples.Low);

            List<Texture> expected = new List<Texture> {
                diffuseSamples.Low,
                bumpSamples.Low,
                emissiveSamples.Low
            };

            return new List<TestCaseData> { new TestCaseData(textures, expected) };
        }

        public static List<TestCaseData> GetHighestAvailableResolutionForEachType_Mix()
        {
            SampleTextures diffuseSamples = new SampleTextures(TextureType.Diffuse);
            SampleTextures bumpSamples = new SampleTextures(TextureType.Bump);
            SampleTextures emissiveSamples = new SampleTextures(TextureType.Emissive);

            List<Texture> textures = new List<Texture>();
            textures.AddRange(new List<Texture> {
                diffuseSamples.Low,
                diffuseSamples.Medium,
                diffuseSamples.High
            });
            textures.AddRange(new List<Texture> {
                bumpSamples.Low,
                bumpSamples.Medium
            });
            textures.Add(emissiveSamples.Low);

            List<Texture> expected = new List<Texture> {
                diffuseSamples.High,
                bumpSamples.Medium,
                emissiveSamples.Low
            };

            return new List<TestCaseData> { new TestCaseData(textures, expected) };
        }

        public static List<TestCaseData> GetHighestAvailableResolutionForEachType_Empty()
        {
            SampleTextures diffuseSamples = new SampleTextures(TextureType.Diffuse);
            SampleTextures bumpSamples = new SampleTextures(TextureType.Bump);
            SampleTextures emissiveSamples = new SampleTextures(TextureType.Emissive);

            List<Texture> textures = new List<Texture>();
            diffuseSamples.High.Path = "";
            bumpSamples.Medium.Path = "";
            emissiveSamples.Low.Path = "";

            textures.AddRange(new List<Texture> {
                diffuseSamples.Low,
                diffuseSamples.Medium,
                diffuseSamples.High
            });
            textures.AddRange(new List<Texture> {
                bumpSamples.Low,
                bumpSamples.Medium
            });
            textures.Add(emissiveSamples.Low);

            List<Texture> expected = new List<Texture> {
                diffuseSamples.Medium,
                bumpSamples.Low
            };

            return new List<TestCaseData> { new TestCaseData(textures, expected) };
        }

        public static List<TestCaseData> IsEquivalent_WhenGivenAList_ReturnsExpected_NullAndEmpty()
        {
            return new List<TestCaseData> { new TestCaseData(new List<Texture>(), null, false) };
        }

        public static List<TestCaseData> IsEquivalent_WhenGivenAList_ReturnsExpected_Contents()
        {

            List<Texture> allCopy = new List<Texture>(All);
            List<Texture> allWithExtra = new List<Texture>(All);

            allWithExtra.Add(new Texture
            {
                QualityEnum = TextureQuality.Medium,
                TypeEnum = TextureType.Opacity,
                Path = "extra texture"
            });

            All.Reverse();

            return new List<TestCaseData> {
                new TestCaseData(All, All, true),
                new TestCaseData(allCopy, allCopy, true),
                new TestCaseData(allCopy, All, true),
                new TestCaseData(All, OnlyLow, false),
                new TestCaseData(All, allWithExtra, false)
            };
        }
    }
}
